fortune-mod/util/strfile.8
.\"	$NetBSD: strfile.8,v 1.3 1995/03/23 08:28:45 cgd Exp $
.\"
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Ken Arnold.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)strfile.8	8.1 (Berkeley) 6/9/93
.\"
.Dd June 9, 1993
.Dt STRFILE 8
.Os BSD 4
.Sh NAME
.Nm strfile ,
.Nm unstr
.Nd "create a random access file for storing strings"
.Sh SYNOPSIS
.Nm strfile
.Op Fl iorsx
.Op Fl c Ar char
.Ar source_file
.Op Ar output_file
.Nm unstr
.Ar source_file
.Sh DESCRIPTION
.Nm Strfile
reads a file containing groups of lines separated by a line containing
a single percent
.Ql \&%
sign and creates a data file which contains
a header structure and a table of file offsets for each group of lines.
This allows random access of the strings.
.Pp
The output file, if not specified on the command line, is named
.Ar source_file Ns Sy .out .
.Pp
The options are as follows:
.Bl -tag -width "-c char"
.It Fl c Ar char
Change the delimiting character from the percent sign to
.Ar char .
.It Fl i
Ignore case when ordering the strings.
.It Fl o
Order the strings in alphabetical order.
The offset table will be sorted in the alphabetical order of the
groups of lines referenced.
Any initial non-alphanumeric characters are ignored.
This option causes the
.Dv STR_ORDERED
bit in the header
.Ar str_flags
field to be set.
.It Fl r
Randomize access to the strings.
Entries in the offset table will be randomly ordered.
This option causes the
.Dv STR_RANDOM
bit in the header
.Ar str_flags
field to be set.
.It Fl s
Run silently; don't give a summary message when finished.
.It Fl x
Note that each alphabetic character in the groups of lines is rotated
13 positions in a simple caesar cypher.
This option causes the
.Dv STR_ROTATED
bit in the header
.Ar str_flags
field to be set.
.El
.Pp
The format of the header is:
.Bd -literal
#define	VERSION	2
uint32_t	str_version;	/* version number */
uint32_t	str_numstr;	/* # of strings in the file */
uint32_t	str_longlen;	/* length of longest string */
uint32_t	str_shortlen;	/* length of shortest string */
#define	STR_RANDOM	0x1	/* randomized pointers */
#define	STR_ORDERED	0x2	/* ordered pointers */
#define	STR_ROTATED	0x4	/* rot-13'd text */
uint32_t	str_flags;	/* bit field for flags */
uint8_t	str_delim;	/* delimiting character */
.Ed
.Pp
All fields are written in network byte order.  Each field is also
written independently so as to avoid structure padding problems on
some architectures.
.Pp
The purpose of
.Nm unstr
is to undo the work of
.Nm strfile .
It prints out the strings contained in the file
.Ar source_file
in the order that they are listed in
the header file
.Ar source_file Ns Pa .dat
to standard output.
It is possible to create sorted versions of input files by using
.Fl o
when
.Nm strfile
is run and then using
.Nm unstr
to dump them out in the table order.
.Sh SEE ALSO
.Xr byteorder 3 ,
.Xr fortune 6
.Sh FILES
.Bl -tag -width strfile.out -compact
.It Pa strfile.out
default output file.
.El
.Sh HISTORY
The
.Nm strfile
utility first appeared in 4.4BSD.


fortune-mod/ChangeLog
(Note: this file has been re-arranged to be in reverse chronological
order, which is The Right Thing for ChangeLogs - DLC)

March 05, 2004 (fortune-mod-1.99.1)

	Most of the changes have occurred at some point in time in the last
	5 years.

	A high number of spelling, punctuation, formatting and grammar
	fixes.

	Internationalisation support.

	New -c option to see which file a fortune came from.

August 29, 1997

  Incorporated a couple of minor changes made in the (old) Debian
  fortune-mod package, including the addition of an extra data file
  called 'cookie'.  Renamed some documentation files, and included some
  install information for non-Linux users.  Added a "-v" option to
  report the program version.

  I intend to submit this distribution to SunSITE RSN.

    -- Dennis L. Clark <dbugger@progsoc.uts.edu.au>

May, 1997

  This release fixes many of the portability problems with the
  fortune-mod program released by Amy Lewis in October, 1995.  The
  previous version had many Linux-isms in it, which left it unworkable
  on any other platform.  This version replaces most of these with more
  standard calls, making it more likely to work under other platforms.
  The Makefiles have been modified so that GNU's gcc and make are no
  longer required: any standard make and ANSI-compatible C compiler
  should work.  Sorry, pre-ANSI compilers are not supported (c'mon, this
  is the 90's, darn it!)

  This version has been tested to work on SunOS 4.1.x as well as Linux.
  All changes made to the sources were as platform-independant as
  possible.  Therefore, no "#ifdef LINUX" or "#ifdef SUNOS4" directives
  appear in the code.  An effect of this is that a number of 'implicit
  declaration' warnings are emitted by gcc under SunOS 4.1.x, but this a
  problem with SunOS's standard headers, not with the program or the
  compiler.  The benefit of this approach is that it eases the work of
  expanding the port to include other platforms. Bug reports and fixes
  for other platforms are most welcome!

  A few "standard" C function calls were replaced with more standard
  counterparts at various points.  Generally, when there was a choice
  between a BSD version of a function and a POSIX version, the POSIX
  version was favoured (even though fortune originated on BSD).  An
  exception to this was the regex functions: either POSIX or BSD
  versions can be used, with selection made via the top-level Makefile.

  Fortune and strfile also compiles on Solaris 2.5, but a discrepency
  between the declaration and implementation of 'struct dirent' on the
  test platform caused fortune to execute incorrectly there.  It is not
  certain whether this is a bug in the header file, the C library, or
  the test platform.

  While the making of this release was not meant to become a bug
  search-and-destroy mission, some bugs were inadvertently discovered
  and fixed. including the known bug of using -a with a file that occurs
  in both the offensive and inoffensive directories.  Fixing this
  particular bug required a way to be able to separately identify two
  fortune files with the same name, with one in the inoffensive
  directory, the other one in the offensive directory.  Now, such a name
  will be taken to be the inoffensive file by default.  However, you can
  now append '-o' to a fortune name, and the '-o' will be removed and
  the offensive directory will be searched.  Thus you can say (assuming
  you use the distributed datfiles):

        fortune 80% politics politics-o

  Which has an 80% change of giving an "inoffensive" political fortune,
  and a 20% change of giving an "offensive" political one.  Note that
  this makes fortune-mod backwards-compatible with BSD fortune, but only
  for users, not fortune database maintainers.

  Of course, this solution only passes the buck: with the above example
  again, if you have a 'politics-o' file in your inoffensive directory,
  you are back to square one.  OTOH, seeing that '-o' originally was
  meant for offensive fortunes, using it for inoffensive ones is simply
  asking for trouble.

  Bug fixed: Fortune's definition of a fortune length (for -s and -l)
  was inconsistent.  Unordered fortunes counted the 2 delimiting
  characters (as hinted in the man page), but sorted or randomized
  fortunes did not.  Now the delimiting characters are _NEVER_ counted,
  so you will always get the length limit you expect.

  Another bug fix: -l and -s can now work together with -m.  Previously
  -l and -s were ignored when -m was in effect.  The new behaviour helps
  me count how many long or short fortunes there are in a file.

  Ansify has been removed from the package, as well as some filter
  scripts from NetBSD that no longer appear useful.  Randstr has been
  kept, but has not been improved in any way.  It at least has a man
  page: maybe somebody will find it useful.

  Some of the documentation (including the man pages) has been improved
  and updated, and some files have been renamed so that the package
  looks less Linux-specific.

    -- Dennis L. Clark <dbugger@progsoc.uts.edu.au>

Late October 1995
  Ansify has been abandoned.  I'm going to distribute this working version
  of fortune, and then see how difficult it would be to add termcap/terminfo
  enhancements to fortune itself--I don't anticipate serious problems, but
  I'd rather go ahead and get this on the net.

  A last-minute change was made to the way that percentages are displayed
  with -f; it is now in the format nnn.nn%.  The reason for this is that
  with the multiplication of small files, fortune -af displayed a large
  number of "0%"s--no worse than the old version, but not helpful.

  The fortunes database was finally cleaned up, and this version is now
  being distributed (at least, I hope it is).  I don't consider the current
  division of fortunes among files absolutely canonical; some are certainly
  in the wrong places.  But things are *better*.

Amy A. Lewis	alewis@email.unc.edu


fortune-mod/util/strfile.c
/*      $NetBSD: strfile.c,v 1.3 1995/03/23 08:28:47 cgd Exp $  */

/*-
 * Copyright (c) 1989, 1993
 *      The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Ken Arnold.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by the University of
 *      California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * Changes, September 1995, to make the damn thing actually sort instead
 * of just pretending.  Amy A. Lewis
 *
 * And lots more.
 *
 * Fixed the special cases of %^J% (an empty fortune), no 'separator' at
 * the end of the file, and a trailing newline at the end of the file, all
 * of which produced total ballsup at one point or another.
 *
 * This included adding a routine to go back and write over the last pointer
 * written or stored, for the case of an empty fortune.
 *
 * unstr also had to be modified (well, for *lots* of reasons, but this was
 * one) to be certain to put the delimiters in the right places.
 */

/*
 *
 #ifndef lint
 static char copyright[] =
 "@(#) Copyright (c) 1989, 1993\n\
 The Regents of the University of California.  All rights reserved.\n";
 #endif / * not lint * /

 #ifndef lint
 #if 0
 static char sccsid[] = "@(#)strfile.c  8.1 (Berkeley) 5/31/93";
 #else
 static char rcsid[] = "$NetBSD: strfile.c,v 1.3 1995/03/23 08:28:47 cgd Exp $";
 #endif
 #endif / * not lint * /
 *
 *I haven't the faintest flipping idea what all that is, so kill the warnings
 */

#include        <stdlib.h>
#include        <unistd.h>
#include        <netinet/in.h>
#include        <sys/param.h>
#include        <stdio.h>
#include        <ctype.h>
#include        <string.h>
#include        <time.h>
#include        "strfile.h"

#ifndef MAXPATHLEN
#define MAXPATHLEN      1024
#endif /* MAXPATHLEN */

/*
 *    This program takes a file composed of strings separated by
 * lines containing only the delimiting character (the default
 * character is '%') and creates another file which consists of a table
 * describing the file (structure from "strfile.h"), a table of seek
 * pointers to the start of the strings, and the strings, each terminated
 * by a null byte.  Usage:
 *
 *      % strfile [-iorsx] [ -cC ] sourcefile [ datafile ]
 *
 *      c - Change delimiting character from '%' to 'C'
 *      s - Silent.  Give no summary of data processed at the end of
 *          the run.
 *      o - order the strings in alphabetic order
 *      i - if ordering, ignore case
 *      r - randomize the order of the strings
 *      x - set rotated bit
 *
 *              Ken Arnold      Sept. 7, 1978 --
 *
 *      Added ordering options.
 *
 * Made ordering options do more than set the bloody flag, September 95 A. Lewis
 *
 * Always make sure that your loop control variables aren't set to bloody
 * *zero* before distributing the bloody code, all right?
 *
 */

#define FALSE   0

#define STORING_PTRS    (Oflag || Rflag)
#define CHUNKSIZE       512

#define ALWAYS  1
#define ALLOC(ptr,sz)   if (ALWAYS) { \
                        if (ptr == NULL) \
                                ptr = malloc((unsigned int) (CHUNKSIZE * sizeof *ptr)); \
                        else if (((sz) + 1) % CHUNKSIZE == 0) \
                                ptr = realloc((void *) ptr, ((unsigned int) ((sz) + CHUNKSIZE) * sizeof *ptr)); \
                        if (ptr == NULL) { \
                                fprintf(stderr, "out of space\n"); \
                                exit(1); \
                        } \
                }

typedef struct
{
    char first;
    int32_t pos;
}
STR;

static char *Infile = NULL,            /* input file name */
  Outfile[MAXPATHLEN] = "",     /* output file name */
  Delimch = '%';                /* delimiting character */

static int Sflag = FALSE;              /* silent run flag */
static int Oflag = FALSE;              /* ordering flag */
static int Iflag = FALSE;              /* ignore case flag */
static int Rflag = FALSE;              /* randomize order flag */
static int Xflag = FALSE;              /* set rotated bit */
static long Num_pts = 0;               /* number of pointers/strings */

static int32_t *Seekpts;

static FILE *Sort_1, *Sort_2;          /* pointers for sorting */

static STRFILE Tbl;                    /* statistics table */

static STR *Firstch;                   /* first chars of each string */

static void __attribute__((noreturn)) usage(void)
{
    fprintf(stderr,
            "strfile [-iorsx] [-c char] sourcefile [datafile]\n");
    exit(1);
}

/*
 *    This routine evaluates arguments from the command line
 */
static void getargs(int argc, char **argv)
{
    int ch;

    while ((ch = getopt(argc, argv, "c:iorsx")) != EOF)
        switch (ch)
          {
          case 'c':             /* new delimiting char */
              Delimch = *optarg;
              if (!isascii(Delimch))
              {
                  printf("bad delimiting character: '\\%o\n'",
                         Delimch);
              }
              break;
          case 'i':             /* ignore case in ordering */
              Iflag++;
              break;
          case 'o':             /* order strings */
              Oflag++;
              break;
          case 'r':             /* randomize pointers */
              Rflag++;
              break;
          case 's':             /* silent */
              Sflag++;
              break;
          case 'x':             /* set the rotated bit */
              Xflag++;
              break;
          case '?':
          default:
              usage();
          }
    argv += optind;

    if (*argv)
    {
        Infile = *argv;
        if (*++argv)
            (void) strcpy(Outfile, *argv);
    }
    if (!Infile)
    {
        puts("No input file name");
        usage();
    }
    if (*Outfile == '\0')
    {
        strcpy(Outfile, Infile);
        strcat(Outfile, ".dat");
    }
}

/*
 * add_offset:
 *      Add an offset to the list, or write it out, as appropriate.
 */
static void add_offset(FILE * fp, int32_t off)
{
    if (!STORING_PTRS)
    {
        uint32_t net;
        net = htonl((uint32_t)off);
        fwrite(&net, 1, sizeof net, fp);
    }
    else
    {
        ALLOC(Seekpts, Num_pts + 1);
        Seekpts[Num_pts] = off;
    }
    Num_pts++;
}

/*
 * fix_last_offset:
 *     Used when we have two separators in a row.
 */
static void fix_last_offset(FILE * fp, int32_t off)
{
    if (!STORING_PTRS)
    {
        uint32_t net = htonl((uint32_t)off);
        fseek(fp, -(long)(sizeof net), SEEK_CUR);
        fwrite(&net, 1, sizeof net, fp);
    }
    else
        Seekpts[Num_pts - 1] = off;
}

/*
 * cmp_str:
 *      Compare two strings in the file
 */
static int cmp_str(const void *v1, const void *v2)
{
    register int c1, c2;
    register int n1, n2;
    register const STR *p1, *p2;

#define SET_N(nf,ch)    (nf = (ch == '\n'))
#define IS_END(ch,nf)   (ch == Delimch && nf)

    p1 = (const STR *) v1;
    p2 = (const STR *) v2;
    c1 = p1->first;
    c2 = p2->first;
    if (c1 != c2)
        return c1 - c2;

    fseek(Sort_1, p1->pos, 0);
    fseek(Sort_2, p2->pos, 0);

    n1 = FALSE;
    n2 = FALSE;
    while (!isalnum(c1 = getc(Sort_1)) && c1 != '\0')
        SET_N(n1, c1);
    while (!isalnum(c2 = getc(Sort_2)) && c2 != '\0')
        SET_N(n2, c2);

    while (!IS_END(c1, n1) && !IS_END(c2, n2))
    {
        if (Iflag)
        {
            if (isupper(c1))
                c1 = tolower(c1);
            if (isupper(c2))
                c2 = tolower(c2);
        }
        if (c1 != c2)
            return c1 - c2;
        SET_N(n1, c1);
        SET_N(n2, c2);
        c1 = getc(Sort_1);
        c2 = getc(Sort_2);
    }
    if (IS_END(c1, n1))
        c1 = 0;
    if (IS_END(c2, n2))
        c2 = 0;
    return c1 - c2;
}

/*
 * do_order:
 *      Order the strings alphabetically (possibly ignoring case).
 */
static void
  do_order(void)
{
    register long i;
    register int32_t *lp;
    register STR *fp;

    Sort_1 = fopen(Infile, "r");
    Sort_2 = fopen(Infile, "r");
    qsort((char *) Firstch, (size_t)( (int) Num_pts - 1), sizeof *Firstch, cmp_str);
/*      i = Tbl.str_numstr;
 * Fucking brilliant.  Tbl.str_numstr was initialized to zero, and is still zero
 */
    i = Num_pts - 1;
    lp = Seekpts;
    fp = Firstch;
    while (i--)
        *lp++ = fp++->pos;
    fclose(Sort_1);
    fclose(Sort_2);
    Tbl.str_flags |= STR_ORDERED;
}

#if 0
static char *
  unctrl(char c)
{
    static char buf[3];

    if (isprint(c))
    {
        buf[0] = c;
        buf[1] = '\0';
    }
    else if (c == 0177)
    {
        buf[0] = '^';
        buf[1] = '?';
    }
    else
    {
        buf[0] = '^';
        buf[1] = c + 'A' - 1;
    }
    return buf;
}
#endif

/*
 * randomize:
 *      Randomize the order of the string table.  We must be careful
 *      not to randomize across delimiter boundaries.  All
 *      randomization is done within each block.
 */
static void randomize(void)
{
    register int cnt, i;
    register int32_t tmp;
    register int32_t *sp;

    srandom((unsigned int) (time((time_t *) NULL) + getpid()));

    Tbl.str_flags |= STR_RANDOM;
/*      cnt = Tbl.str_numstr;
 * See comment above.  Isn't this stuff distributed worldwide?  How embarrassing!
 */
    cnt = (int)Num_pts;

    /*
     * move things around randomly
     */

    for (sp = Seekpts; cnt > 0; cnt--, sp++)
    {
        i = random() % cnt;
        tmp = sp[0];
        sp[0] = sp[i];
        sp[i] = tmp;
    }
}

/*
 * main:
 *      Drive the sucker.  There are two main modes -- either we store
 *      the seek pointers, if the table is to be sorted or randomized,
 *      or we write the pointer directly to the file, if we are to stay
 *      in file order.  If the former, we allocate and re-allocate in
 *      CHUNKSIZE blocks; if the latter, we just write each pointer,
 *      and then seek back to the beginning to write in the table.
 */
int main(int ac, char **av)
{
    register char *sp;
    register FILE *inf, *outf;
    register int32_t last_off, length, pos, *p;
    register int first, cnt;
    register char *nsp;
    register STR *fp;
    static char string[257];

    getargs(ac, av);            /* evalute arguments */
    if ((inf = fopen(Infile, "r")) == NULL)
    {
        perror(Infile);
        exit(1);
    }

    if ((outf = fopen(Outfile, "w")) == NULL)
    {
        perror(Outfile);
        exit(1);
    }
    if (!STORING_PTRS)
        (void) fseek(outf, sizeof Tbl, 0);

    /*
     * Write the strings onto the file
     */

    Tbl.str_longlen = 0;
    Tbl.str_shortlen = (unsigned int) 0xffffffff;
    Tbl.str_delim = (uint8_t)Delimch;
    Tbl.str_version = VERSION;
    first = Oflag;
    add_offset(outf, (int32_t)ftell(inf));
    last_off = 0;
    do
    {
        sp = fgets(string, 256, inf);
        if (sp == NULL || STR_ENDSTRING(sp, Tbl))
        {
            pos = (int32_t)ftell(inf);
            length = pos - last_off - (int32_t)(sp ? strlen(sp) : 0);
            if (!length)
                /* Here's where we go back and fix things, if the
                 * 'fortune' just read was the null string.
                 * We had to make the assignment of last_off slightly
                 * redundant to achieve this.
                 */
            {
                if (pos - last_off == 2)
                    fix_last_offset(outf, pos);
                last_off = pos;
                continue;
            }
            last_off = pos;
            add_offset(outf, pos);
            if ((int)Tbl.str_longlen < length)
                Tbl.str_longlen = (uint32_t)length;
            if ((int)Tbl.str_shortlen > length)
                Tbl.str_shortlen = (uint32_t)length;
            first = Oflag;
        }
        else if (first)
        {
            for (nsp = sp; !isalnum(*nsp); nsp++)
                continue;
            ALLOC(Firstch, Num_pts);
            fp = &Firstch[Num_pts - 1];
            if (Iflag && isupper(*nsp))
                fp->first = (char)tolower(*nsp);
            else
                fp->first = *nsp;
            fp->pos = Seekpts[Num_pts - 1];
            first = FALSE;
        }
    }
    while (sp != NULL);

    /*
     * write the tables in
     */

    fclose(inf);

    if (Oflag)
        do_order();
    else if (Rflag)
        randomize();

    if (Xflag)
        Tbl.str_flags |= STR_ROTATED;

    if (!Sflag)
    {
        printf("\"%s\" created\n", Outfile);
        if (Num_pts == 1)
            puts("There was no string");
        else
        {
            if (Num_pts == 2)
                puts("There was 1 string");
            else
                printf("There were %ld strings\n", Num_pts - 1);
            printf("Longest string: %lu byte%s\n", (unsigned long)(Tbl.str_longlen),
                Tbl.str_longlen == 1 ? "" : "s");
            printf("Shortest string: %lu byte%s\n", (unsigned long)(Tbl.str_shortlen),
                Tbl.str_shortlen == 1 ? "" : "s");
        }
    }

    fseek(outf, (off_t) 0, 0);
    Tbl.str_version = htonl(Tbl.str_version);
    Tbl.str_numstr = htonl((uint32_t)(Num_pts - 1));
    /* Look, Ma!  After using the variable three times, let's store
     * something in it!
     */
    Tbl.str_longlen = htonl(Tbl.str_longlen);
    Tbl.str_shortlen = htonl(Tbl.str_shortlen);
    Tbl.str_flags = htonl(Tbl.str_flags);
    fwrite(&Tbl.str_version,  sizeof Tbl.str_version,  1, outf);
    fwrite(&Tbl.str_numstr,   sizeof Tbl.str_numstr,   1, outf);
    fwrite(&Tbl.str_longlen,  sizeof Tbl.str_longlen,  1, outf);
    fwrite(&Tbl.str_shortlen, sizeof Tbl.str_shortlen, 1, outf);
    fwrite(&Tbl.str_flags,    sizeof Tbl.str_flags,    1, outf);
    fwrite( Tbl.stuff,        sizeof Tbl.stuff,        1, outf);
    if (STORING_PTRS)
    {
        for (p = Seekpts, cnt = (int)Num_pts; cnt--; ++p)
        {
            *p = (int32_t)htonl((uint32_t)*p);
            fwrite(p, sizeof *p, 1, outf);
        }
    }
    fclose(outf);
    exit(0);
}
